Group/Task Mutations

kdipaolo opened the pull request:
- Created mutation for a new task
- Created mutation for an updated task
- Created mutation for a new group
- Created mutation for an updated group

- Created Settings components in individual Group/Task views (WIP)
- Created Alert component for when user  does an action this would pop up confirming the action or communicating the action completed (WIP)

*If testing, i would use responsive mode in inspector to test as a mobile device.  Desktop view has not been optimized yet.

Questions:
- Wanting to make a notification components which shows a success/failure message for a second or two then goes away every time a user does a certain action.
- How would i get the settings list dropdown to animate?

kdipaolo committed:
Comments from meeting

kdipaolo committed:
mutations/queries added for groups/tasks

raygesualdo commented on src/App.js:
```diff
@@ -50,12 +50,15 @@ const client = new ApolloClient({
 
 const AppBackground = withRouter(
   styled.div`
     ${props =>
```
Over >>> on the new file, to keep things clean, you could pull lines 60-63 out into a function. Something like this: 

```javascript
const setBackgroundColor = props => props.location.pathname === '/' ? css`background: linear-gradient(#5A6ED0 0%, #4053AE 100%);` : css`background: #fff;`

// ... 
styled.div`
  ${setBackgroundColor}
  background-repeat: no-repeat;
`
```

raygesualdo commented on src/components/cards/Group.js:
```diff
@@ -1,15 +1,26 @@
+    const { id, name, tasks, created, dueDate } = this.props
     return (
-      <Link to="/group/R3JvdXA6Mg==">
+      <Link to={'/group/' + id}>
```
Don't forget about template string literals. This line is perfectly correct, but it could also be written as

```javascript
<Link to={`/group/${id}`}>
```

raygesualdo commented on src/components/cards/InfoCard.js:
```diff
@@ -2,14 +2,18 @@ import React from 'react'
 import styled from 'styled-components'
 import placeholder from '../../../public/images/placeholder.png'
 import EditIcon from 'react-icons/lib/md/edit'
+import gear from 'react-icons/lib/fa/cog'
```
Good call importing these individually. I didn't in 3Blades/react-frontend and it bloated my production build size by ~250kb.

raygesualdo commented on src/components/cards/InfoCard.js:
```diff
@@ -37,77 +41,68 @@ const Image = styled.img`
     this.state = {
       edit: false,
-      description: null
+      value: null
```
What exactly is this state used for? I see the state being destructured in line 61, but it doesn't seem to be updated anywhere. If this state isn't being updated anywhere, might as well pull it out into a constant and convert this class component to a function.

kdipaolo replied:
I think it was from a previous solution I had.  I went ahead and removed the references to state.

raygesualdo commented on src/components/cards/InfoCard.js:
```diff
@@ -37,77 +41,68 @@ const Image = styled.img`
+`
+
 class InfoCard extends React.Component {
   constructor(props) {
```
Let's talk about the latest Stage 3 public class fields. They allow us to clean up the `constructor` a lot (read: get rid of it).

raygesualdo commented on src/components/cards/InfoCard.js:
```diff
@@ -37,77 +41,68 @@ const Image = styled.img`
+            handleGroupUpdate={this.props.handleGroupUpdate}
+          />}
+        <Info>
+          <Gear onClick={e => this.props.handleStateUpdate(e, 'settings')} />
```
We touched on it last week, but note that inlining functions like this will bring a performance hit because new function is created each time `render()` is called. However, in an app your size, the performance hit will probably be imperceptible.

raygesualdo commented on src/components/cards/InfoCard.js:
```diff
@@ -37,77 +41,68 @@ const Image = styled.img`
+          />}
+        <Info>
+          <Gear onClick={e => this.props.handleStateUpdate(e, 'settings')} />
+          {member && <Image src={placeholder} alt="" />}
```
Way to include empty alt tags! #a11y

raygesualdo commented on src/components/cards/InfoCard.js:
```diff
@@ -37,77 +41,68 @@ const Image = styled.img`
+            {!task &&
+              <Detail>
+                <Highlight>
+                  {tasks.filter(task => task.completed).length}
```
Nice. Love any use of `.filter`, `.map`, and `.reduce`.

raygesualdo commented on src/components/shared/ActionSlide.js:
```diff
@@ -1,46 +1,45 @@
   text-align: center;
-  padding:10% 5%;
+  padding: 10% 5%;
   form {
```
Why are there nested tags inside this s/c? Shouldn't they be their own components?

raygesualdo commented on src/components/shared/ActionSlide.js:
```diff
@@ -61,33 +60,32 @@ const ActionSlideWrapper = styled.div`
         <ButtonWrapper>
-          <Button>+ Add {this.props.type}</Button>
-          <Button onClick={this.props.handleCloseClick}>x Close</Button>
+          <Button onClick={e => handleClose(e)} name="openMenu">
```
Since all that's being done here is calling `handleClose` with the event, it can be more simply written `onClick={handleClose}`.

raygesualdo commented on src/components/shared/ActionSlide.js:
```diff
@@ -61,33 +60,32 @@ const ActionSlideWrapper = styled.div`
+        <p>
+          + Add a New {type}
+        </p>
+        <form onSubmit={e => handleAdd(e, this.refs)}>
```
Let's talk through alternatives to using refs in forms.

raygesualdo commented on src/components/shared/Button.js:
```diff
@@ -23,19 +23,23 @@ function stickyButton() {
-  ${props => props.notification && createColorVariation(props.theme.colors.primaryDark)}
-  ${props => props.secondary && createColorVariation(props.theme.colors.secondary)}
-  ${props => props.sticky && stickyButton()}
+  ${props =>
```
For clarity, it might help to un-nest these rules.

raygesualdo commented on src/components/shared/Message.js:
```diff
@@ -30,12 +30,11 @@ const MessageWrapper = styled.div`
-  ${props => props.response && response()}
-  ${props => props.completed && completed()}
-  ${props => props.rejected && rejected()}
+  ${props => props.response && response()} ${props =>
```
It might help to write the `response` function (and similar functions) as an arrow function, like so:
```javascipt
const response = () => css``
```

raygesualdo commented on src/components/shared/Switcher.js:
```diff
@@ -15,18 +15,19 @@ const SwitcherItem = styled.p`
     return (
       <SwitcherWrapper>
-        {this.props.links.map(link => {
+        {this.props.links.map((link, i) => {
```
When only returning a component inside an arrow function, it can be simplified like so:
```javascript
{this.props.links.map((link, i) => (
  <Component />
))}
```
No need for extra curly braces and a `return` statement.

raygesualdo commented on src/components/views/Group.js:
```diff
@@ -3,93 +3,172 @@ import Button from '../shared/Button'
-    this.handleButtonClick = this.handleButtonClick.bind(this)
+
     this.handleSwitcherClick = this.handleSwitcherClick.bind(this)
+    this.handleGroupUpdate = this.handleGroupUpdate.bind(this)
```
Part of the class fields discussion will include getting rid of `this.methodName.bind(this)` as well 😁

raygesualdo commented on src/components/views/Groups.js:
```diff
@@ -1,16 +1,113 @@
+
+  handleNewGroup(e, refs) {
+    e.preventDefault()
+    this.props.client
```
Should this be handled by a mutation injected as a prop instead of `props.client.mutate()`?

raygesualdo commented on src/styles/theme/global.js:
```diff
@@ -1,6 +1,6 @@
 // 'polished/lib/mixins/normalize';
 import normalizeStyled from 'normalize-styled'
```
I couldn't get this working when I cloned the repo. `npm` couldn't find `normalize-styled`. Polished comes with a normalize function, so `normalize-styled` shouldn't be necessary.

kdipaolo commented:
@raygesualdo Thanks for taking the time to provide feedback.  I love it! Excited to talk through these later today.

raygesualdo commented:
Also, `dateformat` isn't included in the `package.json`.

kdipaolo committed:
Changed links to template literal strings"
"

kdipaolo committed:
removed constructor class syntax from all components

kdipaolo committed:
removed iinline functions

kdipaolo committed:
updated other input specific components to adopt state handling of the input value

kdipaolo committed:
Updated formatting of button component functions

kdipaolo committed:
Updated package.json to remove normalized styles package

kdipaolo committed:
Added delete functionality for groups and tasks

kdipaolo commented:
@raygesualdo I went ahead and fixed most of the issues we talked through last week in this PR and added the ability to delete groups and tasks.  Excited to talk through this today.
